rules/gsuite_activityevent_rules/google_workspace_new_login_ip.py

```diff
@@ -0,0 +1,111 @@
+import json
+import ast
+from datetime import datetime, timedelta
+
+from panther_base_helpers import deep_get
+from panther_oss_helpers import get_string_set, put_string_set
+
+ALERT_CONTEXT_DICTIONARY = {}
+
+# Store the record in Dynamo for each user, for this amount of days
+# If the user does not log in within this period, the record will expire, and the next login will alert
+DYNAMO_CACHE_DAYS = 30
+
+
+def rule(event):
+    # see: https://developers.google.com/admin-sdk/reports/v1/appendix/activity/login#login
+    is_login_event = event.get("type") == 'login' and event.get("name") == 'login_success'
+    if not is_login_event:
+        return False
+
+    ip_address = get_ip_address(event)
+    user_identifier = get_user_identifier(event)
+
+    if not ip_address or not user_identifier:
+        # We can only alert if there was an IP and a user_identifier (email) associated with the login
+        return False
+
+    event_key = get_dynamo_key(user_identifier)
+
+    # name|user_identifier : set(ip_address)
+    user_ip_history_raw = load_from_dynamo(event_key)
+    user_ip_history = list(user_ip_history_raw) # cannot append to a set() or serialize it to json
+
+    ALERT_CONTEXT_DICTIONARY['previous_ips'] = user_ip_history.copy()
+
+    # If no previous login record exists, store the current login and alert
+    no_previous_logins_for_user = not user_ip_history
+
+    # This IP has not logged in before - alert, and then store the IP to prevent subsequent alerts
+    new_ip_detected = ip_address not in user_ip_history
+
+    if no_previous_logins_for_user or new_ip_detected:
```

Comment:
no_previous_logins_for_user may be unnecessary since it will always be a new_ip_detected if there is no previous logins